compiler/test/dotty/tools/dotc/ParallelTesting.scala

```diff
@@ -0,0 +1,1076 @@
+      }
+      else if (shouldFail && !test.didFail) {
+        fail("Run test should have failed, but did not")
+      }
+
+      cleanup()
+    }
+
+    /** Deletes output directories and files */
+    private def cleanup(): this.type = {
+      if (shouldDelete) targets.foreach(t => delete(t.outDir))
+      this
+    }
+
+    /** Copies `file` to `dir` - taking into account if `file` is a directory,
+     *  and if so copying recursively
+     */
+    private def copyToDir(dir: JFile, file: JFile): JFile = {
+      val target = Paths.get(dir.getAbsolutePath, file.getName)
+      Files.copy(file.toPath, target, REPLACE_EXISTING)
+      if (file.isDirectory) file.listFiles.map(copyToDir(target.toFile, _))
+      target.toFile
+    }
+
+    /** Builds a new `CompilationTest` where we have copied the target files to
+     *  the out directory. This is needed for tests that modify the original
+     *  source, such as `-rewrite` tests
+     */
+    def copyToTarget(): CompilationTest = new CompilationTest (
+      targets.map {
+        case target @ ConcurrentCompilationTarget(files, _, outDir) =>
+          target.copy(files = files.map(copyToDir(outDir,_)))
+        case target @ SeparateCompilationTarget(dir, _, outDir) =>
+          target.copy(dir = copyToDir(outDir, dir))
+      },
+      times, shouldDelete, threadLimit, shouldFail
+    )
+
+    /** Builds a `CompilationTest` which performs the compilation `i` times on
+     *  each target
+     */
+    def times(i: Int): CompilationTest =
+      new CompilationTest(targets, i, shouldDelete, threadLimit, shouldFail)
+
+    /** Builds a `Compilationtest` which passes the verbose flag and logs the
+     *  classpath
+     */
+    def verbose: CompilationTest = new CompilationTest(
+      targets.map(t => t.withFlags("-verbose", "-Ylog-classpath")),
+      times, shouldDelete, threadLimit, shouldFail
+    )
+
+    /** Builds a `CompilationTest` which keeps the generated output files
+     *
+     *  This is needed for tests like `tastyBootstrap` which relies on first
+     *  compiling a certain part of the project and then compiling a second
+     *  part which depends on the first
+     */
+    def keepOutput: CompilationTest =
+      new CompilationTest(targets, times, false, threadLimit, shouldFail)
+
+    /** Builds a `CompilationTest` with a limited level of concurrency with
+     *  maximum `i` threads
+     */
+    def limitThreads(i: Int): CompilationTest =
+      new CompilationTest(targets, times, shouldDelete, Some(i), shouldFail)
+
+    /** Builds a `CompilationTest` where the executed test is expected to fail
+     *
+     *  This behaviour is mainly needed for the tests that test the test suite.
+     */
+    def expectFailure: CompilationTest =
+      new CompilationTest(targets, times, shouldDelete, threadLimit, true)
+
+    /** Delete all output files generated by this `CompilationTest` */
+    def delete(): Unit = targets.foreach(t => delete(t.outDir))
```

Comment:
That's what I'm saying: the conditional should check condition and call unconditional one instead of re-implementing it.